src/ApparentHorizons/CMakeLists.txt

```diff
@@ -21,6 +21,7 @@ spectre_target_headers(
   HEADERS
   ComputeHorizonVolumeQuantities.hpp
   ComputeHorizonVolumeQuantities.tpp
+  ComputeVolumeQuantities.hpp
```

Comment:
Could you rename this file? Naming it `ComputeVolumeQuantities` gives the impression that it has to do with the actual computation of volume quantities when in reality the function just gets a tag out of a Variables or a TempBuffer. Maybe `ComputeVolumeQuantitiesDetail.hpp` to signify it's a helper or even `GetTagFromTargetOrTemp.hpp`? Not sure which is better

Reply:
Better yet move the file to something like `DataStructures/TaggedContainers.hpp` as it has absolutely nothing to do with ApparentHorizons.  @nilsdeppe you discussed this with me as a new file for a similar type of function...

Reply:
Yes, this implementation is a basic version of what I have... It should take an arbitrary number of containers that have a tag list as their parameters, then peel off the first container until the right container is found. In the case of a `DataBox` it should use the tag retrieveable lookup. while for other cases it's list_contains. There should also be one version that takes all the containers as `const &` and returns a `const &`, and a second that takes all the containers as `gsl::not_null` and returns a `gsl::not_null`

Reply:
For completeness, here's my solution:
```cpp
 template <typename Tag, typename... TagsOrTagsList, typename... Ts,
           template <class...> class U>
 const typename Tag::type& get(const U<TagsOrTagsList...>& first_vars,
                               const Ts&... vars) {
   if constexpr (tt::is_a_v<db::DataBox, U<TagsOrTagsList...>>) {
     if constexpr (not db::tag_is_retrievable_v<U<TagsOrTagsList...>, Tag> and
                   sizeof...(Ts) != 0) {
       return get<Tag>(vars...);
     } else {
       return get<Tag>(first_vars);
     }
   } else {
     if constexpr (not tmpl::list_contains_v<
                       tmpl::flatten<tmpl::list<TagsOrTagsList...>>, Tag> and
                   sizeof...(Ts) != 0) {
       return get<Tag>(vars...);
     } else {
       return get<Tag>(first_vars);
     }
   }
 }

 template <typename Tag, typename... TagsOrTagsList, typename... Ts,
           template <class...> class U>
 gsl::not_null<typename Tag::type*> get(
     const gsl::not_null<U<TagsOrTagsList...>*> first_vars,
     const gsl::not_null<Ts*>... vars) {
   if constexpr (tt::is_a_v<db::DataBox, U<TagsOrTagsList...>>) {
     if constexpr (not db::tag_is_retrievable_v<U<TagsOrTagsList...>, Tag> and
                   sizeof...(Ts) != 0) {
       return get<Tag>(vars...);
     } else {
       return make_not_null(&get<Tag>(*first_vars));
     }
   } else {
     if constexpr (not tmpl::list_contains_v<
                       tmpl::flatten<tmpl::list<TagsOrTagsList...>>, Tag> and
                   sizeof...(Ts) != 0) {
       return get<Tag>(vars...);
     } else {
       return make_not_null(&get<Tag>(*first_vars));
     }
   }
 }
```
This should work on any tagged container that either has a tags list or a set of tags as template parameter, and also works on a DataBox where there might be base tags. I've only tested this in my production code, not written an explicit test for it yet. I would put it in the file Larry suggested, `DataStructures/TaggedContainers.hpp`